### PR TITLE
Advice on hashing cost verses lifetime of user

### PR DESCRIPTION
**Pull request by mvr320**
This is only a suggestion to the user as we do not know the password length.

### CODE REVIEW

File: webapp/src/Service/CheckConfigService.php [hunk 1]
```diff
@@ -391,6 +392,33 @@ public function checkContestsValidate()
...
+        // Fetch all active and future contests
+        $contests = $this->dj->getCurrentContests(null, true);
+
+        $contestWarning = [];
+        $result = 'O';
+        foreach ($contests as $contest) {
+            $cid = $contest->getCid();
+            $duration = $contest->getEndtime()-$contest->getStarttime();
+            $multiplier = 60*60*24*7;
+            if ($duration/$multiplier>1) {
```
Comment (nickygerritsen) [inline webapp/src/Service/CheckConfigService.php:406]: Maybe `$this->getStartTimeObject()->add(new \DateInterval('P1W')) < $this->getEndTimeObject()` is more readable?

Comment (mvr320) [reply webapp/src/Service/CheckConfigService.php:406]: I had to google this... but yes its more readable.

**Comment by mvr320** [comment]
> 
> 
> Passwords are for users, not contests, so the security depends on how long a user account is active, not how long the contest is active?

Valid point, but I assume that a user is valid from start to end? I would say that its still up to the admin to verify if the warning is applicable for that contest. Are you against this merge or should it be implemented differently?

**Comment by thijskh** [comment]
I'm not enthousiastic because I think it conflates two things: lifetime of a password and contest length, which are only loosly coupled. I think this just confuses and muddies the concepts for users.

What about: read out the cost factor, show a statically configured advice for that factor?
If <= 7 then : Your cost factor is 7, this is appropriate when your user's passwords are in use for a week or less. If they live longer, please raise the cost factor.

**Comment by mvr320** [comment]
> 
> 
> I'm not enthousiastic because I think it conflates two things: lifetime of a password and contest length, which are only loosly coupled. I think this just confuses and muddies the concepts for users.
> 
> What about: read out the cost factor, show a statically configured advice for that factor?
> If <= 7 then : Your cost factor is 7, this is appropriate when your user's passwords are in use for a week or less. If they live longer, please raise the cost factor.

Ok, makes sense and is doable, I will also advice on the password length but I think your suggestion is indeed better.

File: webapp/src/Service/CheckConfigService.php [hunk 2]
```diff
@@ -385,6 +394,21 @@ public function checkHashTime() : array
             'desc' => sprintf('Hashing cost is reasonable (Did %d hashes).', $counter)];
     }
 
+    public function checkHashStrength() : array
+    {
+        if ($this->cryptCost>7) {
```
Comment (nickygerritsen) [inline webapp/src/Service/CheckConfigService.php:399]: Add spaces around `>`

**Review by nickygerritsen — APPROVED** [review]

**Comment by thijskh** [comment]
As it appears the current advice is not complete. As the chosen hash cost may also significantly affect API performance and thus judging througput. I think we need to be explicit about that. It's not only the length of the user account lifetime but also how good you want the judging to perform.

**Comment by mvr320** [comment]
> 
> 
> As it appears the current advice is not complete. As the chosen hash cost may also significantly affect API performance and thus judging througput. I think we need to be explicit about that. It's not only the length of the user account lifetime but also how good you want the judging to perform.

I think we could do something like:
https://symfony.com/doc/current/security/named_encoders.html
assuming that someone makes the password longer, the hash can be less costly for the judgedaemon.

So I would propose to have the password for restapi a bit longer, and the hash maybe less costly?

**Comment by thijskh** [comment]
I've continued the discussion in #936 since that seems more appropriate than a closed PR

**Pull request merged**